trieste/space.py

```diff
@@ -329,3 +347,160 @@ def __mul__(self, other: Box) -> Box:
+        self._subspace_sizes = dict(zip(self._tags, subspace_sizes))
+        self._subspace_starting_indicies = dict(
+            zip(self._tags, tf.cumsum(subspace_sizes, exclusive=True))
+        )
+        self._dimension = tf.reduce_sum(subspace_sizes)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"""TaggedProductSearchSpace(spaces =
+                {[self.get_subspace(tag) for tag in self.subspace_tags]},
+                tags = {self.subspace_tags})
+                """
+
+    @property
+    def subspace_tags(self) -> tuple[str, ...]:
+        """Return the names of the subspaces contained in this product space."""
+        return self._tags
+
+    @property
+    def dimension(self) -> TensorType:
+        """The input dimensionality of this product search space."""
+        return self._dimension
+
+    def get_subspace(self, tag: str) -> SearchSpace:
+        """Return the domain of a particular subspace."""
+        return self._spaces[tag]
+
+    def get_subspace_component(self, tag: str, values: TensorType) -> TensorType:
+        """
+        Returns the components of ``values`` lying in a particular subspace.
+
+        :param value: Points from the :class:`DecomposableSearchSpace` of shape [N,D].
+        :return: The context components of ``values`` of shape [M, D], where
+            M is the dimensionality of the specified subspace.
+        """
+
+        starting_index_of_subspace = self._subspace_starting_indicies[tag]
+        ending_index_of_subspace = starting_index_of_subspace + self._subspace_sizes[tag]
+        return values[:, starting_index_of_subspace:ending_index_of_subspace]
+
+    def __contains__(self, value: TensorType) -> bool | TensorType:
+        """
+        Return `True` if ``value`` is a member of this search space, else `False`. A point is a
+        member if each of its subspace components lie in each subspace.
+
+        Recall that individual points in product spaces are represented with their inputs in the
+        same order as specified when initializing the space.
+
+        :param value: A point to check for membership of this :class:`SearchSpace`.
+        :return: `True` if ``value`` is a member of this search space, else `False`. May return a
+            scalar boolean `TensorType` instead of the `bool` itself.
+        :raise ValueError (or tf.errors.InvalidArgumentError): If ``value`` has a different
+            dimensionality from the search space.
+        """
+
+        tf.debugging.assert_equal(
+            tf.shape(value),
+            self.dimension,
+            message="value must have same dimensionality as search space",
+        )
+        value = tf.expand_dims(value, 0)
+        in_each_subspace = [
+            self._spaces[tag].__contains__(self.get_subspace_component(tag, value)[0, :])
+            for tag in self._tags
+        ]
+        return tf.math.reduce_all(in_each_subspace)
```

Comment:
NIT: There's a reduce_all in the top level tf module, as well as tf.math, so this would be ever so slightly more succinct

```suggestion
        return tf.reduce_all(in_each_subspace)
```

Reply:
Never really worked out why they include it twice